### ISSUE TITLE
Is it possible to sandbox a vm?

### ISSUE THREAD

**Issue by ShaneFarrell** [pypyjs/pypyjs#158]
I'm wondering if it would be possible to run two pypyjs vms on a page, and have one of them be sandboxed so that it can't interact with the page and can only communicate through stdout / stderr.

That way I could run my app in python and be able to run untrusted user code. I don't mind running it in a web worker.

**Comment by jedie** [comment]
JavaScript is generally a kind of a sandbox.
I see no problem in running untrusted user code,yet. Everything is running in the client browser and nothing on the server...

**Comment by ShaneFarrell** [comment]
Yeah, but the main thing I want to do is stop untrusted code doing things like redirecting and modifying the DOM (and freezing the page, but I assume using a webworker will fix that).

**Comment by jedie** [comment]
Modifying the DOM would be made with the "js" modul. So IMHO, you can just replace the "js" modul with a restricted one...

But why? Manipulating the DOM will be not saved.

**Comment by ShaneFarrell** [comment]
I'm trying to make a game with a level editor that allows players to write custom code for objects. There will be an ingame way for players to play each other's levels. I essentially want to make sure the players can't write problematic object code that could do things like phishing attacks or redirecting to porn sites etc.

**Comment by rfk** [comment]
Whoops, just realized I hadn't replied to this.

I think this should be fairly easy to do, particularly if you can make a custom build of pypyjs without the `js` module.  All emscripten-compiled code runs in a kind of sandbox anyway, the only way it can interact with the host environment is via functions that you pass into it explicitly when you set it up.

**Comment by perkinslr** [comment]
As far as running two pypyjs vms on a page, one inside a webworker, it works quite well.  You'll still want to deny access to the js module, since even in a webworker you can open websockets and make xmlhttp requests.  My fork is out of date, but has some scripts that may help you get it up and running in a webworker.  It runs the main script in a webworker and uses multiprocessing to communicate with the drawing thread.
